src/datasets/load.py

```diff
@@ -133,6 +135,79 @@ def import_main_class(module_path, dataset=True) -> Optional[Union[Type[DatasetB
     return module_main_cls
 
 
+class _InitializeConfiguredDatasetBuilder:
+    """
+    From https://stackoverflow.com/questions/4647566/pickle-a-dynamically-parameterized-sub-class
+    See also ConfiguredDatasetBuilder.__reduce__
+
+    When called with the param value as the only argument, returns an
+    un-initialized instance of the parameterized class. Subsequent __setstate__
+    will be called by pickle.
+    """
+
+    def __call__(self, builder_cls, metadata_configs, default_config_name, name):
+        # make a simple object which has no complex __init__ (this one will do)
+        obj = _InitializeConfiguredDatasetBuilder()
+        obj.__class__ = configure_builder_class(
+            builder_cls, metadata_configs, default_config_name=default_config_name, dataset_name=name
+        )
+        return obj
+
+
+def configure_builder_class(
+    builder_cls: Type[DatasetBuilder],
+    builder_configs: List[BuilderConfig],
+    default_config_name: Optional[str],
+    dataset_name: str,
+) -> Type[DatasetBuilder]:
+    """
+    Dynamically create a builder class with custom builder configs parsed from README.md file,
+    i.e. set BUILDER_CONFIGS class variable of a builder class to custom configs list.
+    """
+
+    class ConfiguredDatasetBuilder(builder_cls):
+        BUILDER_CONFIGS = builder_configs
+        DEFAULT_CONFIG_NAME = default_config_name
+
+        __module__ = builder_cls.__module__  # so that the actual packaged builder can be imported
+
+        def __reduce__(self):  # to make dynamically created class pickable, see _InitializeParameterizedDatasetBuilder
+            parent_builder_cls = self.__class__.__mro__[1]
+            return (
+                _InitializeConfiguredDatasetBuilder(),
+                (
+                    parent_builder_cls,
+                    self.BUILDER_CONFIGS,
+                    self.DEFAULT_CONFIG_NAME,
+                    self.dataset_name,
+                ),
+                self.__dict__.copy(),
+            )
+
+    ConfiguredDatasetBuilder.__name__ = (
+        f"{builder_cls.__name__.lower().capitalize()}{snakecase_to_camelcase(dataset_name)}"
+    )
+    ConfiguredDatasetBuilder.__qualname__ = (
+        f"{builder_cls.__name__.lower().capitalize()}{snakecase_to_camelcase(dataset_name)}"
+    )
+
+    return ConfiguredDatasetBuilder
```

Comment:
Thank you. I see...
I think we should implement a test in our CI for this pickle/unpickle, so that it fails if the builder class is not properly created inside a function scope.

Reply:
It is implicitly tested in tests for parallel loading ([test_load_hub_dataset_without_script_with_metadata_config_in_parallel()](https://github.com/huggingface/datasets/pull/5331/files#diff-39f545cd21f1ca75281b3f357c296be83c964764f99dc24888f9e1e3f8fa05c6)), but I can definitely add a specific one for it.

Reply:
I tested this proposal with `type` in our CI, and `test_load_hub_dataset_without_script_with_metadata_config_in_parallel` passed: https://github.com/huggingface/datasets/actions/runs/5067215702/jobs/9098858661

Reply:
This is the branch where I tried this approach (the CI integration tests are unrelated): https://github.com/huggingface/datasets/commits/ci-tmp-albert-config-params

Reply:
Thanks!!! After your clear explanation, I understand better now.